### ISSUE TITLE
[Idea] add a high performance path with binary transfer and direct access to an ArrayBuffer

### ISSUE THREAD

**Issue by 0x53A** [Zaid-Ajaj/Fable.Remoting#116]
_This is a bit different from the current goal of Fable.Remoting, and may be out-of-scope for you. But I like the programming model of Fable.Remoting. So I wanted to at least ask you for your opinion._

Even though the below suggestion is already detailed, I'm not too married to it. I would just really like _any_ way to transfer big(ish) blobs without the (de)serialization overhead.

----------------------

I have a scenario where I need to transfer big objects.

In the first prototype I serialized them to base64 strings then transfered them over Fable.Remoting. In that case, the underlying web request took ~2 seconds, but the F.R Api call took 20 seconds. So 18 seconds were spent just parsing and deserializing big base64 encoded strings.

Currently I have a secondary Giraffe Webhandler which returns the binary data, and on the client side I do a ``fetch`` and access the ArrayBuffer directly.

With this and other improvements I was able to reduce total time for request+processing from over a minute to 10 seconds.

While not a big issue currently, this has a few downsides:

1) I have two seperate APIs for no reason
2) Duplicated authentication
3) no easy input-parameter handling (currently I have just an int-array in the body, so not much of an issue)

---------

I would like to unify this again with the rest of my API and would propose to add a binary fast-path to Fable.Remoting:

```f#
// Add this to Fable.Remoting
type BinaryBuffer() =

    // Fable Client Side
    member this.GetArrayBuffer() : Async<Fable.Import.JS.ArrayBuffer> = // ...

    // .NET ServerSide
    static member FromStream(s:System.IO.Stream) : BinaryBuffer = // ...
    static member FromArray(a:byte array) : BinaryBuffer = // ...
```

My API record would then look like this:

```f#
type IRemotingApi = {
    // typical request, serialized as json
    GetSomeData: Async<MyData>
    // fast binary response, no serialization & deserialization involved
    Load3DModelsForIDs: int array -> Async<BinaryBuffer>
}
```

On the server, I would create a buffer from either an array or a stream, and on the client I would use it like this:

```f#
let ids = [| 1;2;3 |]
let! response = api.Load3DModelsForIDs(ids)
let! buffer = response.GetArrayBuffer()

// now I can work with the array buffer

// maybe the buffer represents a list of vertices (floats) for my 3d model
let typedArray = JS.Float32Array.Create(buffer)
```

__Downsides:__

* I lose the strong typing, the blob can represent anything
* basically two almost completely different models munched together into one library

__Upsides:__

* only one unified api
* coherent input parameter handling

__Open questions:__

* error handling? Currently I use ``Result<'TOk, 'TErr>`` for all my responses, but the BinaryBuffer would need to be the sole direct return value - it couldn't be nested into a response.

**Comment by Zaid-Ajaj** [comment]
This is an interesting idea, though normally I would suggest to keep doing what you are already doing: a seperate handler for very specific non-Remoting work such as this one. But if we look at the issue from the perspective of "Byte deserialization is slow for big arrays" then I can give it a try:

I am thinking about an example implementation and my only conditions for is feature are:
 - To be hidden away without extra config (i.e. no `BinaryBuffer`) and use just `byte[]`
 - To have limited impact (in code size) because I don't want to restructure the repository just to make this work.

> I will need to investigate the impact on the code base 

The first thought I got was to make a special case for handling routes where the output type is `Async<byte[]>` these special routes will not get serialized as `Base64` string and deserialized with SimpleJson. Instead, we can write the bytes to the response stream directly and use `arraybuffer` from the client to convert it to typed array:

```js
var xhr = new XMLHttpRequest();
xhr.open("POST", "/api/Load3DModels"); 
xhr.responseType = "arraybuffer"; 
xhr.onreadystatechange = function() { 
    if (xhr.readyState === 4) {
        var arrayBuffer = xhr.response;
        var output = U8IntArray(arrayBuffer);
    }
}

xhr.send(JSON.stringify(yourIDs);
```
Here we are skipping the whole "loop through the base64 to create the U8IntArray" which is supposedly the reason for the slow deserialization, am I correct? 

Also the serialization of the input parameters (the `int list`) is kept the same as other routes but only the deserialization is different. 

What do you think?

**Comment by 0x53A** [comment]
You are right, I was thinking too complicated again. 🤦‍♂️ 

Just special-casing ``byte array`` should work, Fable already represents them as ``UInt8Array``.

-----------------------------

> Here we are skipping the whole "loop through the base64 to create the U8IntArray" which is supposedly the reason for the slow deserialization, am I correct?

Tbh I didn't profile the browser, and in my first prototype it was a little bit more complicated - I had a ``Map<int, byte [] []>``, so it wasn't just one big byte array but many medium sized. But conversions to/from arrays are unexpectedly expensive in JS so I would like to avoid all conversions if possible.

----------------

> Also the serialization of the input parameters (the int list) is kept the same as other routes but only the deserialization is different.

I don't have such a scenario yet, but could the exact same optimization also be applied to input parameters? If I have exactly one parameter of ``byte array``, skip serialization and write it directly?

----------------

> What do you think?

sounds perfect
